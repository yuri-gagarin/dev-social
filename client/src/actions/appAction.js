import axios from "axios";

export const test = () => dispatch => {
  setTimeout(function() {
    dispatch({
      type: "TEST",
      payload: "Now Tested"
    });
  }, 1000);
};

export const cancelTest = () => {
  return {
    type: "CANCEL_TEST",
    payload: "Cancelld the test"
  };
};


