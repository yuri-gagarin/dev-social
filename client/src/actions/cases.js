export const TEST = "TEST";
export const CANCEL_TEST = "CANCEL_TEST";
export const LIST_ERRORS = "LIST_ERRORS";

export const REGISTER = "REGISTER";
export const LOGIN  = "LOGIN";
export const SET_USER = "SET_USER";
export const LOGOUT = "LOGOUT";


