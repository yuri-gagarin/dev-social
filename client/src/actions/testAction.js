
export const testAction = () => {
  return (dispatch) => {
    dispatch({
      type: "TEST_ACTION",
      payload: "result of the test action"
    });
  }
};

