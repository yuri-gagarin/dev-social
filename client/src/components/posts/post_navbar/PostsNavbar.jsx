import React, {Component} from "react";
import PropTypes from "prop-types";
import {postSearchOptions as options} from "../../../redux/searchOptions.js";
import { Menu, Dropdown, Icon } from "semantic-ui-react";
import styles from "../../../assets/stylesheets/posts/post.scss";

class PostsNavbar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filter: options.filter.new,
      from: options.time.day,
      filterBarEnabled: true,
      timeBarEnabled: false,
    }
  }

  handleSortClick = (e) => {
    const value = e.target.getAttribute("data-value");
    switch (value) {
      case options.filter.new:
        this.setState({filter: value, from: options.time.none, timeBarEnabled: false}, () => {
          const fetchOptions = {
            filter: this.state.filter,
            from: this.state.time,
            limit: this.state.limit,
          };
          this.props.fetchPosts(fetchOptions)
        });
        break;
      case options.filter.trending:
        this.setState({filter: value, from: options.time.none, timeBarEnabled: false}, () => {
          const fetchOptions = {
            filter: this.state.filter,
            from: this.state.time,
            limit: this.state.limit,
          };
          this.props.fetchPosts(fetchOptions);
        });
        break;
      case options.filter.controversial:
        this.setState({filter: options.filter.controversial, from: options.time.day, timeBarEnabled: true},  () => {
          const fetchOptions = {
            filter: this.state.filter,
            from: this.state.time,
            limit: this.state.limit,
          };
          this.props.fetchPosts(fetchOptions);
        });
        break;
      case options.filter.discussed:
        this.setState({filter: options.filter.discussed, from: options.time.day, timeBarEnabled: true}, () => {
          const fetchOptions = {
            fitler: this.state.filter,
            from: this.state.time,
            limit: this.state.limit,
          };
          this.props.fetchPosts(fetchOptions);
        });
        break;
      default:
        this.setState({filter: options.filter.new, from: options.time.none, timeBarEnabled: false}, () => {
          const fetchOptions = {
            filter: this.state.filter,
            from: this.state.time,
            limit: this.state.limit,
          };
          this.props.fetchPosts(fetchOptions);
        });
    }
  }

  handleTimeClick = (e) => {
    const value = e.target.getAttribute("data-value");
    this.setState({from: value} , () => {
      //convert the time filter
      const fetchOptions = {
        filter: this.state.filter,
        from: this.state.from,
        limit: this.state.limit,
      };
      this.props.fetchPosts(fetchOptions);
    });
  }

  render() {
    return (
      <Menu className={styles.postsNavbar} style={{marginTop: "50px", width: "100vw"}} data-test={"posts-navbar"}>
        <Dropdown className={styles.sortPostsDropdown} data-test="nav-filter"
          text="Filter Posts"
          icon="caret square down"
          pointing
          floating
          labeled
          button >
          <Dropdown.Menu className={styles.menuDropdown} style={{margin: "0em !important"}}>
            <Dropdown.Header icon='filter' content='Filter by type' />
            <Dropdown.Divider />
            <Dropdown.Item onClick={this.handleSortClick} data-value={options.filter.new}>
              <div className={styles.filterOption}>
                <i aria-hidden="true" className="icon feed"></i>
                <span>{options.filter.new}</span>
              </div>
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleSortClick} data-value={options.filter.trending}>
              <div className={styles.filterOption}>
                <i aria-hidden="true" className="icon bullhorn"></i>
                <span>{options.filter.trending}</span>
              </div>
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleSortClick} data-value={options.filter.controversial}>
              <div className={styles.filterOption}>
                <i aria-hidden="true" className="icon fire"></i>
                <span>{options.filter.controversial}</span>
              </div>
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleSortClick} data-value={options.filter.discussed}>
              <div className={styles.filterOption}>
                <i aria-hidden="true" className="icon comments"></i>
                <span>{options.filter.discussed}</span>
              </div>
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>
        <div className={styles.sortValue}>{this.state.filter}</div>
        <Dropdown className={styles.sortPostsTime} data-test="post-time-filter"
          text="Time"
          icon="time"
          pointing
          floating
          labeled
          disabled={!this.state.timeBarEnabled}
          button >
          <Dropdown.Menu>
            <Dropdown.Header content="Filter Time"/>
            <Dropdown.Divider />
            <Dropdown.Item onClick={this.handleTimeClick} data-value={options.time.day}>
              {options.time.day}
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleTimeClick} data-value={options.time.week}>
              {options.time.week}
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleTimeClick} data-value={options.time.month}>
              {options.time.month}
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleTimeClick} data-value={options.time.year}>
              {options.time.year}
            </Dropdown.Item>
            <Dropdown.Item onClick={this.handleTimeClick} data-value={options.time.allTime}>
              {options.time.allTime}
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>
        <div className={styles.sortValue}>{this.state.from}</div>
      </Menu>
    );
  }
};

PostsNavbar.propTypes = {
  authstate: PropTypes.object,
  fetchPosts: PropTypes.func.isRequired,
};

export default PostsNavbar;