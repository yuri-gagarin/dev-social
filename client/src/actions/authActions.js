import axios from "axios";
import {REGISTER, LIST_ERRORS, LOGIN, LOGOUT} from "../actions/cases.js";

export const registerUserTest = (newUserData) => {
  console.log(newUserData);

  return function(dispatch) {
    const config = {
      method: "POST",
      url: "/api/users/register_test",
      data: newUserData
    };
    axios(config)
      .then((response) => {
        console.log(response);
        dispatch({
          type: REGISTER,
          payload: response
        });
      })
      .catch((error) => {
        console.log("error");
        console.log(error.response);
        dispatch({
          type: LIST_ERRORS,
          payload: {
            statusCode: error.response.status,
            data: error.response.data
          }
        });
      });
  }
};

export const registerUser = (newUserData) => {
  return function(dispatch) {
    const config = {
      method: "POST",
      url: "/api/users/register",
      data: newUserData
    };
    axios(config)
      .then((response) => {
        dispatch({
          type: REGISTER,
          payload:  {
            status: response.status,
            data: response.data
          }
        });
      })
      .catch((error) => {
        dispatch({
          type: LIST_ERRORS,
          payload: {
            statusCode: error.response.status,
            data: error.response.data
          }
        });
      });
  };
};

export const loginUser = (clientData) => {
  return function(dispatch) {
    const config = {
      method: "POST",
      url: "/api/users/login",
      data: clientData,
    };
    axios(config)
      .then((response) => {
        dispatch({
          type: LOGIN,
          payload: {
            statusCode: response.status,
            data: response.data,
          }
        });
        return Promise.resolve();
      })
      .catch((error) => {
        dispatch({
          type: LIST_ERRORS,
          payload: {
            message: error.response.data.message,
            statusText: error.response.statusText,
            status: error.response.status,
          }
        });
      });
  }
}