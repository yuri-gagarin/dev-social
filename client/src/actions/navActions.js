import {OPEN_DASH, CLOSE_DASH} from "./cases.js";

export const openDashboard = () => {
  return function(dispatch) {
    //should come from the server depending on user type
    //api call
    const dashData = {
      user: "user",
      type: "admin",
      options: {

      }
    };

    dispatch({
      type: OPEN_DASH,
      payload: dashData,
    })
  }
};

export const closeDashboard = () => {
  return {
    type: CLOSE_DASH, 
    payload: null,
  }
};

