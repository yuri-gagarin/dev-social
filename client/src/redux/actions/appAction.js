import axios from "axios";
import pluralize from "pluralize";

import {LIST_ERRORS, MODEL_ERROR, FETCH_DATA, FETCH_POSTS} from "../cases.js";
import isEmpty from "../../../../helpers/validators/isEmpty.js";

/**
 * Capitalizes a string. 
 * @throws {TypeError} Throws a  TypeError if argument is not a string.
 * @param {String} string A string to be capitalized.
 * @returns {String} A capitalized String.
 */
const capitalize = (string) => {
  if (typeof string !== "string") {
    throw new TypeError("Expected fist argument to be a {String}");
  }
  return string.charAt(0).toUpperCase() + string.slice(1);
};
const getModelAction = (dataModel) => {
  switch (dataModel) {
    case "posts":
      return FETCH_POSTS;
    default:
      return MODEL_ERROR;
  }
};

export const test = () => {
  return (dispatch) => {
    setTimeout(function() {
      dispatch({
        type: "TEST",
        payload: "Now Tested"
      });
    }, 1000);
  };
};

export const cancelTest = () => {
  return (dispatch) => {
    setTimeout(function() {
      dispatch({
        type: "CANCEL_TEST",
        payload: "Now Test Cancelled"
      });
    }, 1000);
  };
};

export const fetchData = (options) => {
  const dataModel = options.model;
  const modelName = dataModel ? capitalize(pluralize.singular(dataModel)) : null;
  const dispatchType = getModelAction(modelName);

  return function(dispatch) {
    const error = {};
    if (typeof options !== "object") {
      error.status = 400;
      error.statusText = "Invalid Client Input";
      error.message = "If this error persists please send our team a message";
    }
    if (dispatchType === MODEL_ERROR) {
      error.status = 400;
      error.statusText = "Invalid Input";
      error.message = "Cannot resolve database";
    }
    if(!isEmpty(error)) {
      dispatch({
        type: LIST_ERRORS,
        payload: error,
      })
    }
    else {
      axios({
        method: "get",
        url: options.appRoute,
      })
        .then((response) => {
          const data = response[dataModel];
          console.log(data)
          dispatch({
            type: dispatchType,
            payload: data,
          })
        })
        .catch((error) => {
          console.log(error);
        });
    }
  }
} 
