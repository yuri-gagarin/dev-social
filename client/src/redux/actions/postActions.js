import {FETCH_POSTS,FETCH_TRENDING_POSTS, LIKE_POST, UNLIKE_POST, LIST_ERRORS} from "../cases.js";
import {trimString} from "../../helpers/rendering/displayHelpers.js";
import axios from "axios";

export const fetchPosts = (options={}) => {
  const sortOption = options.sortOption || "new";
  const fetchLimit = options.fetchLimit || 10;
  return function(dispatch) {
    axios({
      method: "get",
      url: `/api/posts?q=${sortOption}`,
      data: {
        fetchLimit: fetchLimit,
      }
    })
    .then((response) => {
      const posts = response.data.posts.map((post) => {
        return {...post, text: trimString(post.text, 100)};
      });
      const postState = {
        message: response.data.message,
        posts: posts,
      };
      dispatch({
        type: FETCH_POSTS,
        payload: postState,
      });
    })
    .catch((error) => {
      dispatch({
        type: LIST_ERRORS,
        payload: error,
      })
    });
  }
};

export const fetchTrendingPosts = () => {
  return function(dispatch) {
    axios({
      method: "get",
      url: `/api/posts?q=trending&l=5`,
    })
    .then((response) => {
      const trendingPosts = response.data.posts.map((post) => {
        return {
          _id: post._id,
          title: post.title,
          author: post.author,
          likeCount: post.likeCount,
        };
      });
      const trendingPostsState = {
        message: response.data.message,
        trendingPosts: trendingPosts,
      };
      dispatch({
        type: FETCH_TRENDING_POSTS,
        payload: trendingPostsState,
      });
    })
    .catch((error) => {
      dispatch({
        type: LIST_ERRORS,
        payload: error
      });
    });
  }
};

export const likePost = (postId, userId) => {
  return function(dispatch) {
    
    ///api actions

    dispatch({
      type: LIKE_POST,
      payload: 0,
    })
  };
};

export const unlikePost = (postId, userId) => {
  return function(dispatch) {
    //api actions
    dispatch({
      type: UNLIKE_POST,
      payload: 0,
    });
  };
};


//downvotes later??
export const downvotePost = (postdId, userId) => {

};
export const removeDownvote = (postId, userId) => {

};
