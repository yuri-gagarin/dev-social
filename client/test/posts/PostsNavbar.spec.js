import React from "react";
import PostsNavbar from "../../src/components/posts/post_navbar/PostsNavbar";
import configureMockStore from  "redux-mock-store";
import thunk from "redux-thunk";
import * as postActions from "../../src/redux/actions/postActions";
import fetchMock from "fetch-mock";

import renderer from "react-test-renderer";

describe("PostNavbar test", function() {
  const component = renderer.create(<PostsNavbar/>);
  let tree = component.toJSON();
  test("Component renders", () => {
    expect(tree).toMatchSnapshot();
  });

  describe("PostNavbar actions", () => {
    afterEach(() => {
      fetchMock.restore();
    });

    it("Successfully fetches Post(s)", () => {
      fetchMock.getOnce("/api/posts", {

      })

    })
  })
})