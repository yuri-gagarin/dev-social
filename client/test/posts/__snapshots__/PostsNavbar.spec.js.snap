// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostNavbar test Component renders 1`] = `
<div
  className="ui postsNavbar menu"
  style={
    Object {
      "marginTop": "50px",
      "width": "100vw",
    }
  }
>
  <div
    aria-expanded={false}
    className="ui button floating labeled pointing dropdown sortPostsDropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="listbox"
    tabIndex={0}
  >
    <div
      aria-atomic={true}
      aria-live="polite"
      className="text"
      role="alert"
    >
      Filter Posts
    </div>
    <i
      aria-hidden="true"
      className="caret square down icon"
      onClick={[Function]}
    />
    <div
      className="menu transition menuDropdown"
      style={
        Object {
          "margin": "0em !important",
        }
      }
    >
      <div
        className="header"
      >
        <i
          aria-hidden="true"
          className="filter icon"
          onClick={[Function]}
        />
        Filter by type
      </div>
      <div
        className="divider"
      />
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        <i
          aria-hidden="true"
          className="bullhorn icon right floated"
          onClick={[Function]}
        />
        New
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        <i
          aria-hidden="true"
          className="heartbeat icon right floated"
          onClick={[Function]}
        />
        Trending
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        <i
          aria-hidden="true"
          className="fire icon right floated"
          onClick={[Function]}
        />
        Heated
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        <i
          aria-hidden="true"
          className="conversation icon right floated"
          onClick={[Function]}
        />
        Discussed
      </div>
    </div>
  </div>
  <div
    className="sortValue"
  >
    New
  </div>
  <div
    aria-expanded={false}
    className="ui button floating labeled pointing dropdown sortPostsTime"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="listbox"
    tabIndex={0}
  >
    <div
      aria-atomic={true}
      aria-live="polite"
      className="text"
      role="alert"
    >
      Time
    </div>
    <i
      aria-hidden="true"
      className="time icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
    >
      <div
        className="header"
      >
        Filter Time
      </div>
      <div
        className="divider"
      />
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        24 Hours
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        7 Days
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        30 Days
      </div>
      <div
        className="item"
        onClick={[Function]}
        role="option"
      >
        All Time
      </div>
    </div>
  </div>
  <div
    className="sortValue"
  />
</div>
`;
